src/Microsoft.ML.Core/Utilities/ThreadUtils.cs

```diff
@@ -4,44 +4,154 @@
+                        // This will wait for up to the idle time before giving up and exiting.
+                        while (TryDequeue(out (Delegate, object, TaskCompletionSource<bool>) item))
+                        {
+                            try
+                            {
+                                if (item.Item1 is Action<object> pts)
+                                {
+                                    pts(item.Item2);
+                                }
+                                else
+                                {
+                                    ((Action)item.Item1)();
+                                }
+                            }
+                            finally
+                            {
+                                item.Item3.SetResult(true);
+                            }
+                        }
+                    });
+                    t.IsBackground = true;
+                    t.Start();
+                }
+
+                bool Enqueue((Delegate, object, TaskCompletionSource<bool>) item)
+                {
+                    // Enqueue the work.  If there are currently fewer threads waiting
+                    // for work than there are work items in the queue, create another
+                    // thread.  This is a heuristic, in that we might end up creating
+                    // more threads than are truly needed, but this whole type is being
+                    // used to replace a previous solution where every work item created
+                    // its own thread, so this is an improvement regardless of any
+                    // such inefficiencies.
+                    lock (_queue)
+                    {
+                        _queue.Enqueue(item);
+
+                        if (_queue.Count <= _availableThreads)
+                        {
+                            Monitor.Pulse(_queue);
+                            return true;
+                        }
+
+                        return false;
+                    }
+                }
+
+                bool TryDequeue(out (Delegate, object, TaskCompletionSource<bool>) item)
+                {
+                    // Dequeues the next item if one is available.  Before checking,
+                    // the available thread count is increased, so that enqueuers can
+                    // see how many threads are currently waiting, with the count
+                    // decreased after.  Each time it waits, it'll wait for at most
+                    // the idle timeout before giving up.
+                    lock (_queue)
+                    {
+                        _availableThreads++;
+                        try
+                        {
+                            while (_queue.Count == 0)
+                            {
+                                if (!Monitor.Wait(_queue, IdleMilliseconds))
```

Comment:
>Monitor.Wait(_queue [](start = 37, length = 19)

So, I have not actually used `Monitor` directly too often, so see if I understand. You above `lock (_queue)`. Then here, you call `Monitor.Wait`, which when called, *releases* the thread, but then once it returns, the lock is re-established. And then you return from this, whereupon the `finally` block runs and `_availableThreads` is decremented still in the critical section. But if the exception is thrown (in the case where the lock was not reacquired at all), we *still* enter the finally section to decrement, even though we must be out  of the critical section in that specific exception-thrown scenario? Is that right and intentional?

Reply:
Can you clarify which exception? I don't think there should be an exception thrown here?

Reply:
Sure. So, let me just clarify, that in the absence of direct experience with this method I am going by [the documentation](https://docs.microsoft.com/en-us/dotnet/api/system.threading.monitor.wait?view=netstandard-2.0). So maybe I am way off base.

Nonetheless, that documentation describes two scenarios where it returns `true` and `false` (in both cases the lock is re-established) and a third possibility which is that it does not return at all, and an exception is thrown. It looks to me like the increment/decrement of this counter relies upon the lock being re-established in order to be valid.

Perhaps we have the attitude that, if an exception is thrown, we're totally screwed anyway. I might accept that logic, but, if that *was* in fact the thinking, I and others might benefit from it being made explicit. (But then why bother decrementing at all?)

Reply:
I think an exception would only occur if there was a bug in the code, we don't _usually_ code defensively around our own code failing for our own bugs (?). If you're asking because there's a `finally` - I agree, I don't know why there's a finally, since if there is an exception here, it will not be handled, and the process will terminate moments after...

Reply:
OK. I just wasn't sure if the only conceivable reason for this was a bug in the code being called. Is that true? That does sort of bring up the question though, what happens *if* an exception is thrown.

Sometimes we throw exceptions due to reasons that have nothing whatever to do with bugs in our code -- e.g., the user invokes something with bad input data, we might conceivably throw in one of these workers. Have we considered that this is possible?

I'd be less concerned if this wasn't an API, but it is.

Reply:
> So, I have not actually used Monitor directly too often, so see if I understand. You above lock (_queue). Then here, you call Monitor.Wait, which when called, releases the _queue lock, but then once it returns (if it returns!), the lock is re-established.

Yes.

> And then you return from this, whereupon the finally block runs and _availableThreads is decremented still in the critical section.

Correct.

> But if the exception is thrown (in the case where the lock was not reacquired at all), we still enter the finally section to decrement, even though we must be out of the critical section in that specific exception-thrown scenario? Is that right and intentional?

The lock will always be reacquired as part of Wait returning/throwing.  You see it documented otherwise somewhere?

> Can you clarify which exception? I don't think there should be an exception thrown here?

Since this is being used correctly (e.g. the lock is held by this thread when the method is called, the argument object isn't null, the timeout is valid, etc.), the only exceptions that might be thrown are a ThreadAbortException or a ThreadInterruptedException.  The former won't be thrown on core, and in practice neither of them would really be thrown since no one is calling .Abort() or .Interrupt() on the Thread object (the former could be thrown as well if this were all in an AppDomain that was getting torn down, again not on core).  So in reality the try/finally is unlikely to matter, I'm just in the habit of it for things like this.  I can remove it if you prefer.

Reply:
That's good enough for me thanks. The odd specificity of the documentation "when it returns" is what made me suspicious that something else might happen if it were somehow interrupted. But if it's not so that's great.